larray/core/session.py

```diff
@@ -1393,87 +1394,85 @@ def display(k, v, is_metadata=False):
+    def __init__(self, axes):
+        if not all([isinstance(axis, (basestring, Axis)) for axis in axes]):
+            raise TypeError('ArrayDef only accepts string or Axis objects')
+        self.axes = axes
 
-    The reason of the second and third restrictions is to ensure the definition
-    of any variable or parameter to be constant throughout the whole code.
-    For example, a user don't need to remember that a new label has been added to
-    an axis of a given array somewhere earlier in the code (in a part of the model
-    written by a colleague). Therefore, these restrictions reduces the risk to deal
-    with unexpected error messages (like 'Incompatible Axes') and make it easier to
-    work in team.
-
-    Parameters
-    ----------
-    filepath: str, optional
-        Path where items have been saved. This can be either the path to a single file, a path to
-        a directory containing .csv files or a pattern representing several .csv files.
-    meta : list of pairs or dict or OrderedDict or Metadata, optional
-        Metadata (title, description, author, creation_date, ...) associated with the array.
-        Keys must be strings. Values must be of type string, int, float, date, time or datetime.
-
-    Notes
-    -----
-    The 'autocomplete' is a feature in which a software predicts the rest of a variable or function
-    name after a user typed the first letters. This feature allows users to use longer but meaningful
-    variable or function names (like 'population_be' instead of 'pbe') and to avoid creating an unwanted
-    new variable by misspelling the name of a given variable (e.g. typing 'poplation = something'
-    ('population' without u) will create a new variable instead of modifying the 'population' variable).
 
+# XXX: the name of the class below is not really important since it will serve as base
+#      for the LazySession and users are expected to inherit from the LazySession when
+#      they define their own classes
+class TypedSession(Session):
+    """
     Examples
     --------
-    >>> class ModelVariables(FrozenSession):
-    ...     LAST_AGE = 120
-    ...     FIRST_OBS_YEAR = 1991
-    ...     LAST_PROJ_YEAR = 2070
-    ...     AGE = Axis('age=0..{}'.format(LAST_AGE))
-    ...     GENDER = Axis('gender=male,female')
-    ...     TIME = Axis('time={}..{}'.format(FIRST_OBS_YEAR, LAST_PROJ_YEAR))
-    ...     CHILDREN = AGE[0:17]
-    ...     ELDERS = AGE[65:]
-    ...     population = zeros((AGE, GENDER, TIME))
-    ...     births = zeros((AGE, GENDER, TIME))
-    ...     deaths = zeros((AGE, GENDER, TIME))
-    >>> m = ModelVariables()
-    >>> m.names    # doctest: +NORMALIZE_WHITESPACE
-    ['AGE', 'CHILDREN', 'ELDERS', 'FIRST_OBS_YEAR', 'GENDER', 'LAST_AGE', 'LAST_PROJ_YEAR', 'TIME', 'births',
-    'deaths', 'population']
+    Content of file 'model_variables.py'
+
+    >>> # ==== MODEL VARIABLES ====
+    >>> class ModelVariables(TypedSession):
+    ...     FIRST_OBS_YEAR = int
```

Comment:
Forget this. I have looked into it, but it seems like the typing module is not flexible enough to type multi-dimensional arrays yet (see https://github.com/python/typing/issues/516). We could invent our own syntax using annotations, and would cause no runtime problem but would be incompatible with any type checker out there, including mypy and PyCharm, so there is no point. Let's go for types and "array defs" as class variables like above.

Reply:
Thanks for looked into it. I didn't try to inherit/use a dataclass because I thought that dataclass was limited to Python 3.7 or higher only.

Reply:
dataclasses are indeed python 3.7+, but (class) variables annotations (PEP 526) are python3.6+.
dataclasses are based on those (and this is what I actually wanted to try using -- sorry for using the wrong word -- those were synonyms in my head when they should not). Either way, this is not ready yet and I am unsure it will ever be flexible enough to be able to provide custom validation functions in Python (to validate that an array actually has the correct axes).